plugin/forwardcrd/README.md

```diff
@@ -0,0 +1,173 @@
+# forwardcrd
+
+## Name
+
+*forwardcrd* - enables proxying DNS messages to upstream resolvers by reading
+CRDs from a Kubernetes cluster
+
+## Description
+
+The *forwardcrd* plugin is used to dynamically configure stub-domains by
+reading a `DNSZone` CRD within a Kubernetes cluster.
+
+See [Configuring Private DNS Zones and Upstream Nameservers in
+Kubernetes](https://kubernetes.io/blog/2017/04/configuring-private-dns-zones-upstream-nameservers-kubernetes/)
+for a description of stub-domains within Kubernetes.
+
+This plugin can only be used once per Server Block.
+
+## Security
+
+This plugin gives users of Kubernetes another avenue of modifying the CoreDNS
+server other than the `coredns` configmap. Therefore, it is important that you
+limit the RBAC and the `namespace` the plugin reads from to reduce the surface
+area a malicious actor can use. Ideally, the level of access to create `DNSZone`
+resources is at the same level as the access to the `coredns` configmap.
+
+## Syntax
+
+~~~
+forwardcrd [ZONES...]
+~~~
+
+With only the plugin specified, the *forwardcrd* plugin will default to the
+zone specified in the server's block. It will allow any `DNSZone` resource that
+matches or includes the zone as a suffix. If **ZONES** are specified it allows
+any zone listed as a suffix.
+
+```
+forwardcrd [ZONES...] {
+  endpoint URL
+  tls CERT KEY CACERT
+  kubeconfig KUBECONFIG [CONTEXT]
+  namespace NAMESPACE
+}
+```
+
+* `endpoint` specifies the **URL** for a remote k8s API endpoint.  If omitted,
+  it will connect to k8s in-cluster using the cluster service account.
+* `tls` **CERT** **KEY** **CACERT** are the TLS cert, key and the CA cert file
+  names for remote k8s connection.  This option is ignored if connecting
+  in-cluster (i.e. endpoint is not specified).
+* `kubeconfig` **KUBECONFIG [CONTEXT]** authenticates the connection to a remote
+  k8s cluster using a kubeconfig file.  **[CONTEXT]** is optional, if not set,
+  then the current context specified in kubeconfig will be used.  It supports
+  TLS, username and password, or token-based authentication.  This option is
+  ignored if connecting in-cluster (i.e., the endpoint is not specified).
+* `namespace` **NAMESPACE** only reads `DNSZone` resources from the namespace
+  listed. If this option is omitted then it will read from all namespaces.
+  **Note**: It is recommended to limit the namespace (e.g to `kube-system`)
+  because this can be potentially misused. It is ideal to keep the level of
+  write access similar to the `coredns` configmap in the `kube-system`
+  namespace.
+
+## Ready
+
+This plugin reports readiness to the ready plugin. This will happen after it has
+synced to the Kubernetes API.
+
```

Comment:
Should add a section (here or elsewhere) describing how this works with and differs from the Corefile serverblock defined stub domains, and a forward plugin in the same block that covers the following points...
* when DNSZone CRDs and serverblock defined stub domains are used, domains defined in the Corefile take precedence (in the event of zone overlap). e.g. if the domain `example.com` is defined in the corefile as a stub domain, and a DNSZone CRD record defined for `sub.example.com`, then `sub.example.com` would get forwarded to the upstream defined in the corefile, not the DNSZone.
* metrics difference: DNSZone metrics all labeled in single zone (the zone of the enclosing serverblock).
* When using forwardcrd and forward in the same serverblock, DNSZones take precedence over forward plugin defined in the same serverblock. e.g. if a DNSZone is defined for `.`, then no queries would be forwarded to the upstream defined the forward plugin of the same serverblock.

precedence is serverblock -> forwardcrd -> forward

Reply:
Makes sense. Added to the readme.